core/templates/dev/head/domain/topic/TopicObjectFactory.js

```diff
@@ -0,0 +1,266 @@
+    };
+
+    Topic.prototype.getAdditionalStoryIds = function() {
+      return this._additionalStoryIds.slice();
+    };
+
+    Topic.prototype.addUncategorizedSkillId = function(skillId) {
+      if (this._uncategorizedSkillIds.indexOf(skillId) === -1) {
+        var isSkillPresentInSubtopics = false;
+        for (var i = 0; i < this._subtopics.length; i++) {
+          var skillIds = this._subtopics[i].getSkillIds();
+          if (skillIds.indexOf(skillId) !== -1) {
+            isSkillPresentInSubtopics = true;
+            break;
+          }
+        }
+        if (!isSkillPresentInSubtopics) {
+          this._uncategorizedSkillIds.push(skillId);
+          return true;
+        }
+        return false;
+      }
+      return false;
+    };
+
+    Topic.prototype.removeUncategorizedSkillId = function(skillId) {
+      var index = this._uncategorizedSkillIds.indexOf(skillId);
+      if (index > -1) {
+        this._uncategorizedSkillIds.splice(index, 1);
+      }
+    };
+
+    Topic.prototype.clearUncategorizedSkillIds = function() {
+      this._uncategorizedSkillIds.length = 0;
+    };
+
+    Topic.prototype.getUncategorizedSkillIds = function() {
+      return this._uncategorizedSkillIds.slice();
+    };
+
+    // Reassigns all values within this topic to match the existing
+    // topic. This is performed as a deep copy such that none of the
+    // internal, bindable objects are changed within this topic.
+    Topic.prototype.copyFromTopic = function(otherTopic) {
+      this._id = otherTopic.getId();
+      this.setName(otherTopic.getName());
+      this.setDescription(otherTopic.getDescription());
+      this.setLanguageCode(otherTopic.getLanguageCode());
+      this._version = otherTopic.getVersion();
+      this.clearAdditionalStoryIds();
+      this.clearCanonicalStoryIds();
+      this.clearUncategorizedSkillIds();
+      this.clearSubtopics();
+
+      var canonicalStoryIds = otherTopic.getCanonicalStoryIds();
+      for (var i = 0; i < canonicalStoryIds.length; i++) {
+        this.addCanonicalStoryId(angular.copy(canonicalStoryIds[i]));
+      }
+
+      var additionalStoryIds = otherTopic.getAdditionalStoryIds();
+      for (var i = 0; i < additionalStoryIds.length; i++) {
+        this.addAdditionalStoryId(angular.copy(additionalStoryIds[i]));
+      }
+
+      var uncategorizedSkillIds = otherTopic.getUncategorizedSkillIds();
+      for (var i = 0; i < uncategorizedSkillIds.length; i++) {
+        this.addUncategorizedSkillId(angular.copy(uncategorizedSkillIds[i]));
+      }
+    };
+
+    // Static class methods. Note that "this" is not available in static
+    // contexts. This function takes a JSON object which represents a backend
+    // topic python dict.
+    Topic.create = function(topicBackendObject) {
+      return new Topic(topicBackendObject);
+    };
+
+    // Create a new, empty topic. This is not guaranteed to pass validation
+    // tests.
+    Topic.createEmptyTopic = function() {
```

Comment:
This could be called initially, when the user clicks create topic for the first time.

Reply:
Can you tell me more about that?

When a user creates a new topic, it wouldn't be from this page. It would be from the topics and skills dashboard, or perhaps somewhere else. But I would have thought that clicking that button would just send a request to the backend and redirect to the topic editor page. It wouldn't create a topic object in the frontend that is used for later editing. So I still don't really see where this would be used.

Am I missing something?

Reply:
Yeah, the topic would be created in the backend and then sent back, will remove this.